hsds/attr_sn.py

```diff
@@ -384,28 +337,318 @@ async def PUT_Attribute(request):
+    verifyRoot(domain_json)
+    
+    req_obj_id = request.match_info.get("id")
+    if not req_obj_id:
+        req_obj_id = domain_json["root"]
+    kwargs = {"obj_id": req_obj_id, "bucket": bucket}
+    attr_items = await _getAttributesFromRequest(request, body, **kwargs)
+    
+    if attr_items:
+        log.debug(f"PUT Attribute {len(attr_items)} attibutes to add")
+    else:
+        log.debug("no attributes defined yet")
+
+    # next, sort out where these attributes are going to
+    
+    obj_ids = {}
+    if "obj_ids" in body:
+        body_ids = body["obj_ids"]
+        if isinstance(body_ids, list):
+            # multi cast the attributes
+            if not attr_items:
+                msg = "no attributes provided"
+                log.warn(msg)
+                raise HTTPBadRequest(reason=msg)
+            else:
+                for obj_id in body_ids:
+                    if not isValidUuid(obj_id):
+                        msg = f"Invalid object id: {obj_id}"
+                        log.warn(msg)
+                        raise HTTPBadRequest(reason=msg)
+                    obj_ids[obj_id] = attr_items
+
+                msg = f"{len(attr_items)} attributes will be multicast to "
+                msg += f"{len(obj_ids)} objects"
+                log.info(msg)
+        elif isinstance(body_ids, dict):
+            if attr_items:
+                msg = "attributes defined outside the obj_ids dict"
+                log.warn(msg)
+                raise HTTPBadRequest(reason=msg)
+            else:
+                for obj_id in body_ids:
+                    if not isValidUuid(obj_id):
+                        msg = f"Invalid object id: {obj_id}"
+                        log.warn(msg)
+                        raise HTTPBadRequest(reason=msg)
+                id_json = body_ids[obj_id]
+                kwargs = {"obj_id": obj_id, "bucket": bucket}
+                obj_items = _getAttributesFromRequest(request, id_json, **kwargs)
+                if obj_items:
+                    obj_ids[obj_id] = obj_items
+            
+                # write different attributes to different objects
+                msg = f"put attributes over {len(obj_ids)} objects"
+        else:
+            msg = f"unexpected type for obj_ids: {type(obj_ids)}"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+    else:
+        # get the object id from the request
+        obj_id = request.match_info.get("id")
+        if not obj_id:
+            msg = "Missing object id"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+        obj_ids[obj_id] = attr_items  # make it look like a list for consistency      
+
+    log.debug(f"obj_ids: {obj_ids}")
+
+    # TBD - verify that the obj_id belongs to the given domain
+    await validateAction(app, domain, req_obj_id, username, "create")
+
+    count = len(obj_ids)
+    if count == 0:
+        msg = "no obj_ids defined"
+        log.warn(f"PUT_Attributes: {msg}")
```

Comment:
This seems to indicate that `PUT_Attributes` is plural with respect to acting on multiple attribute parent objects, but the `PUT_Attributes` in the data node acted on only a single attribute parent object. 

Is this comment saying that, if `count == 1`, the `PUT_Attributes` in the service node acts as the singular `PUT_Attribute`? That doesn't seem accurate, since it could write multiple attributes to a single attribute parent object in one invocation.